Don't replace syntax when output has shadowed ids

AlexKnauth opened the pull request:
```racket
#lang racket/base

(define (and x y) (list x y))

(define (member? e l)
  (if (member e l)
      #t
      #f))
```
should not be refactored to `(and (member e l) #t)` because `and` is shadowed.

jackfirth commented on default-recommendations/conditional-shortcuts-test.rkt:
```diff
@@ -267,3 +267,9 @@ test: "cond with nested cond in last clause without else can't be flattened"
        [else
         (displayln "else")])]))
 ------------------------------
+
+test: "`if-else-false-to-and` is not refactorable when `and` is shadowed"
+------------------------------
+(define (and x y) (list x y))
+(if 'a (println "true branch") #f)
+------------------------------
```
Can you add more tests, ideally with cases that you know will generate replacements that use things like `ORIGINAL-SPLICE` and `NEWLINE`?

AlexKnauth replied:
The existing tests generate replacements with those identifiers mixed into them, which is why I added that in the first place. I don't fully understand why though.

AlexKnauth replied:
Are the new tests I added enough?

jackfirth replied:
Yup. I just haven't had much chance to work on Resyntax lately.

AlexKnauth commented:
I have added more tests. However, I am surprised that the final test I added seems to work, when I expected my changes to be insufficient to catch it:
```racket
test: "shadowed `define-syntax-rule`: single-clause syntax-rules macro"
------------------------------
(define-syntax define-syntax-rule
  (syntax-rules ()
    [(define-syntax-rule new old)
     (define-syntax new
       (syntax-rules ()
         [(new . args) (old . args)]))]))
(define-syntax my-or
  (syntax-rules ()
    [(my-or a b)
     (let ([tmp a])
       (if a a b))]))
------------------------------
```

AlexKnauth commented:
After adding 2 more shadowed-output tests I have a guess about what's going on. I had been confused because it seemed to me to have information about the "future" expansion of the syntax object that it's looking at. But then I realized the way `resyntax` uses `current-expand-observe observe-event!`, it actually fully expands the program before doing any analyses, not just the "partial expansion" that I had imagined in my head from seeing un-expanded syntax objects. Since expansion continues after the expansion observe event, the expander state contains the information it needs to detect these conflicts.

AlexKnauth commented:
> Since expansion continues after the expand-observe event, the expander state contains the information it needs to detect these conflicts.

This same principle also allows it to detect the let-to-define tricky case

jackfirth commented on private/source.rkt:
```diff
@@ -82,20 +93,33 @@
+      [(_ _) (void)])
     (parameterize ([current-expand-observe observe-event!])
       (expand stx))
+    (define (scopes-by-location stx)
+      (define loc (syntax-source-location stx))
+      (hash-ref visits-by-location loc (λ () (hash-ref others-by-location loc #f))))
```
Instead of storing a function in the source code analysis results, I think it might be clearer to store (an immutable copy of) the hash table itself. That way, it's a bit easier to understand what data is being captured and represented by this part of the analysis results.

AlexKnauth replied:
What's wrong with a function? The places that use `scopes-by-location` in other files would not benefit from having a hash table instead. When it's encapsulated in a function, the implementation details of how the data is stored are not their concern.

If you're worried about clarity and understanding, documenting the signature and purpose of the function should accomplish that better than turning it into a hash table, in my opinion.

jackfirth replied:
Functions can do anything. They can have side effects, be nondeterministic, be unexpectedly expensive to compute, etc. If the purpose of this component of the analysis is to map source locations to identifiers with all their scopes added, just give people that map. Documenting the signature and purpose doesn't sway me either since you can also document the signature and purpose of the hash map.

AlexKnauth replied:
I have turned it into a hash table

jackfirth commented on private/syntax-replacement.rkt:
```diff
@@ -295,6 +301,35 @@
+;; ignore-id? : (Syntax -> Syntax) (Listof Id) -> Id -> Boolean
+(define ((ignore-id? scope ignore) id)
+  (or (member id ignore free-identifier=?)
+      (bound-identifier=? id (scope id 'remove))))
```
What's this check for?

jackfirth replied:
Oh I see - you're checking that the identifier doesn't have that scope, right?

jackfirth commented on private/syntax-replacement.rkt:
```diff
@@ -295,6 +301,35 @@
+;; stx-e : Stx -> Any
+(define (stx-e stx) (if (syntax? stx) (syntax-e stx) stx))
+
+;; stx-free-ids=?/only : (Id -> Boolean) -> Stx Stx -> Boolean
+(define (stx-free-ids=?/ignore ignore?)
+  (define (stx-free-ids=? a b)
+    (cond
+      [(and (identifier? a) (identifier? b))
+       (or (free-identifier=? a b) (and (ignore? a) (ignore? b)))]
+      [else
+       (equal?/recur (stx-e a) (stx-e b) stx-free-ids=?)]))
+  stx-free-ids=?)
```
I prefer to avoid relying on `equal?/recur`, it's a little too brittle in my opinion. There's an `in-syntax-identifiers` function I wrote in `resyntax/default-recommendations/private/syntax-identifier-sets` that you can use to implement this logic instead:

```suggestion
;; stx-free-ids=?/ignore : (Id -> Boolean) -> Stx Stx -> Boolean
(define ((stx-free-ids=?/ignore ignore?) stx1 stx2)
  (for/and ([id1 (in-syntax-identifiers stx1)]
            [id2 (in-syntax-identifiers stx2)])
    (or (and (ignore? id1) (ignore? id2)) (free-identifier=? id1 id2))))
```

This also removes the need for recursion. In fact, it gets simple enough that it might be worth just inlining it if it's only used in `syntax-replacement-preserves-free-identifiers?`. Inlining it and the `ignore-id?` predicate would make the purpose of this check easier to understand.

AlexKnauth replied:
I'd rather not use two `in-syntax-identifiers` sequences in parallel on two syntax objects that just so "happen" to have the same tree structure. But this does give me a better idea: I can avoid constructing the second syntax object and just traverse the identifiers in the first!

jackfirth commented on private/syntax-replacement.rkt:
```diff
@@ -295,6 +301,35 @@
+    [(syntax-replacement #:original-syntax orig
+                         #:new-syntax new
+                         #:introduction-scope intro)
+     ((stx-free-ids=?/ignore
+       (ignore-id? intro (list #'SPACE #'NEWLINE #'ORIGINAL-GAP #'ORIGINAL-SPLICE)))
```
Does `stx-free-ids=?/ignore` need to be curried? Why not pass in all the arguments at once?

jackfirth commented on private/syntax-replacement.rkt:
```diff
@@ -12,11 +12,15 @@
-   (-> #:original-syntax (and/c syntax? syntax-original?) #:new-syntax syntax? syntax-replacement?)]
+   (-> #:original-syntax (and/c syntax? syntax-original?)
+       #:new-syntax syntax?
+       #:introduction-scope (->* (syntax?) ((or/c 'flip 'add 'remove)) syntax?)
```
I really wish there was a `scope?` data type exposed from the standard library that wrapped these introducers so we didn't need complex contracts just to create and pass around scopes. Maybe someday someone will get around to adding a `racket/scope` module to the standard library or something.

jackfirth commented on private/syntax-replacement.rkt:
```diff
@@ -295,6 +301,35 @@
+     ((stx-free-ids=?/ignore
+       (ignore-id? intro (list #'SPACE #'NEWLINE #'ORIGINAL-GAP #'ORIGINAL-SPLICE)))
+      new
+      (datum->syntax orig (syntax->datum new)))]))
```
How does this work, exactly? You're checking that all of the identifiers are `free-identifier=?` to themselves when they have all their scopes removed and replaced with just the scopes from the original syntax object, right?

jackfirth reviewed (changes requested):
The tests look great. I finally got around to looking through the actual logic.

jackfirth commented on private/source.rkt:
```diff
@@ -82,20 +93,33 @@
+    (define/match (observe-event! sig val)
+      [('visit val)
+       (add-original-location! visits-by-location val)]
+      [('letX-renames (list-rest trans-idss _ val-idss _))
```
So, what _is_ a `'letX-renames` event? If you reverse engineered it by looking at some other code somewhere, could you add a comment with a link to that code so I can figure it out too?

AlexKnauth commented on private/source.rkt:
```diff
@@ -82,20 +93,33 @@
+    (define/guard (add-original-location! hsh stx)
+      (guard (and (syntax? stx) (syntax-original? stx)) else (void))
+      (define loc (syntax-source-location stx))
+      (guard (hash-has-key? hsh loc) then (void))
+      (hash-set! hsh loc stx))
```
Would it be better as `(hash-ref! hsh loc stx)` with no `guard` or `unless` at all?

AlexKnauth commented on private/source.rkt:
```diff
@@ -52,10 +56,17 @@
 
 
-(define-record-type source-code-analysis (code visited-forms))
+(define-record-type source-code-analysis (code visited-forms scopes-by-location))
```
I added a comment above the record type:

 * `scopes-by-location`: `(-> Syntax (U #f Syntax))`, attempts to find the syntax object with a matching source location that has the scopes from its surrounding context properly added

For example, in `(let ([x a]) b)`, the expander expands it to `(let-values ([(x) a]) b)` and adds `letX-renames` scopes to `x` and `b`. The `scopes-by-location` function produces the versions of `x` and `b` with those scopes.

AlexKnauth commented on private/source.rkt:
```diff
@@ -82,20 +93,33 @@
-      (hash-set! visits-by-location loc val))
+    (define others-by-location (make-hash))
+    (define/guard (add-original-location! hsh stx)
+      (guard (and (syntax? stx) (syntax-original? stx)) else (void))
```
I shortened the `add-original-location!` function to eliminate the `guard` forms and use `hash-ref!` instead.

AlexKnauth commented on private/source.rkt:
```diff
@@ -82,20 +93,33 @@
+       (for* ([trans-ids (in-list trans-idss)] [trans-id (in-list trans-ids)])
+         (add-original-location! others-by-location trans-id))
+       (for* ([val-ids (in-list val-idss)] [val-id (in-list val-ids)])
+         (add-original-location! others-by-location val-id))]
```
Thanks, I've made this change and also added comments about the expander using `trans-idss` for `let-syntaxes` and `val-idss` for `let-values`.
